ast/compile.go

```diff
@@ -1538,6 +1623,52 @@ func (c *Compiler) rewriteExprTerms() {
 	}
 }
 
+func (c *Compiler) rewriteRuleHeadRefs() {
+	f := newEqualityFactory(c.localvargen)
+	for _, name := range c.sorted {
+		WalkRules(c.Modules[name], func(rule *Rule) bool {
+
+			ref := rule.Head.Ref()
+			// NOTE(sr): We're backfilling Refs here -- all parser code paths would have them, but
+			//           it's possible to construct Module{} instances from Golang code, so we need
+			//           to accommodate for that, too.
+			if len(rule.Head.Reference) == 0 {
+				rule.Head.Reference = ref
+			}
```

Comment:
Initially, I had forgotten about this: rego modules can come into the system via code, not just from the output of the parser -- therefore, we need to deal with rules whose Head doesn't have a ref, _even though_ rule Heads coming out of the parser always have it.

That's what happens here. It's also the reason why we differentiate `head.Reference` and `head.Ref()`. `head.Ref()` will _always_ be there since it backfills automatically, `head.Reference` will always be there _in the compiler stages_, since it's inserted early in this stage.

We're also dealing with this in JSON marshalling/unmarshaling. If `ref` is in the JSON, it becomes set accordingly; but anything marshalled to JSON will have a `ref`.